### ISSUE TITLE
Simulation time passing within external

### ISSUE THREAD

**Issue by jeroenvanstraten** [cocotb/cocotb#645]
tl;dr: **Is it intended behavior for an external thread to run in parallel to the simulator thread after the first function call, and thus cause non-deterministic simulation time?**

Within our company (Technolution), we use a generated Python module to abstract generated memory-mapped registers in the programmable logic domain to Python properties. This Python module simply provides a read and write hook that needs to be overridden by the underlying interface, for instance an Ethernet communication interface with an FPGA. We would like to use the exact same generated modules within cocotb. Since the module has no awareness of cocotb, all register accesses must be wrapped in an external.

What we've noticed is that tests which do this do not have a deterministic simulation time, which seems to boil down to simulation time being lost between two consecutive function calls in an external, and/or from the last function call to the firing of the external's join trigger. This can be reproduced with the following code snippet in `tests/test_cases/test_external/text_external.py`:

```python
@cocotb.function
def yield_to_next_clock(dut):
    yield RisingEdge(dut.clk)

def test_ext_time_between_functions(dut, clk_period):
    dut._log.info("Clock edge %d" % 0)
    yield_to_next_clock(dut)
    base_time = get_sim_time('ps')
    dut._log.info("Base time = %d ps", base_time)
    for i in range(1, 501):
        dut._log.info("Clock edge %d" % i)
        yield_to_next_clock(dut)
        for _ in range(10):
            _t = get_sim_time('ps')
            dut._log.info("Time reported = %d ps", _t)
            if _t != base_time + clk_period * i:
                raise TestFailure("Time reported does not match expected time %d ps != %d ps" %
                                  (_t, base_time + clk_period * i))
    dut._log.info("external function has ended")

@cocotb.test(expect_fail=False)
def test_time_between_functions(dut):
    """Test that the simulation time does not advance in the external between
    function calls, i.e. that it is synchronized to the simulator"""
    period = 100
    clk_gen = cocotb.fork(Clock(dut.clk, period).start())
    yield Timer(10, 'ns')
    yield cocotb.external(test_ext_time_between_functions)(dut, period)
```

(See also [this commit](https://github.com/Technolution/cocotb/commit/a7a2933f5c8f02558c83c26d21889b4261c094b1))

I would like to work towards a solution for this issue, but I'm not sure if this behavior is actually intended or not, because I see use cases for the current behavior as well. For instance, an external that connects Ethernet signals from the DUT to a virtual Ethernet interface of the host operating system, allowing software running in real-time to connect to the DUT as if it were a real (albeit much slower) piece of hardware. In this case, you probably don't want for instance the lack of packets arriving from the host from halting simulation time.

If the sketched behavior is considered to be a bug, the culprit is [`function.__call__()`](https://github.com/potentialventures/cocotb/blob/6de0683870d58154bc41b4eea43ffc5b0ebd60f6/cocotb/decorators.py#L289). It calls [`waiter.thread_resume()`](https://github.com/potentialventures/cocotb/blob/6de0683870d58154bc41b4eea43ffc5b0ebd60f6/cocotb/scheduler.py#L116) to resume the external without somehow blocking the cocotb main thread/simulator. I spent some time trying to work out a solution to this, but could not think of a good way; the scheduler would need to know that it needs to block itself and wait for the external again when 
the [`execute_function()`](https://github.com/potentialventures/cocotb/blob/6de0683870d58154bc41b4eea43ffc5b0ebd60f6/cocotb/decorators.py#L292) coroutine in [`function.__call__()`](https://github.com/potentialventures/cocotb/blob/6de0683870d58154bc41b4eea43ffc5b0ebd60f6/cocotb/decorators.py#L289) returns/raises `StopIteration`, but since this appears to be just any regular coroutine to the scheduler it currently has no way of knowing this. One could of course create yet another special exception to inform it, but I would argue that this is rather convoluted...

If the sketched behavior is intended, I would like to suggest adding a new kind of decorator that satisfies the generalization of our use case: calling non-cocotb-aware Python modules with callbacks that need to be able to advance simulation time, in such a way that simulation time remains deterministic. I made a proof of concept implementation for this [here](https://github.com/Technolution/cocotb/commit/45324b4e30b1e1b9aae7986a680ed2a5ab18c77e). I've coined them "processes" as an allegory to VHDL processes, but the name is more of a placeholder than anything else. The idea is that such "processes" can call external code (like an external), and the callbacks use `process_yield()` instead of the yield statement. More generally, any `@coroutine` can be replaced with `@process`, using `process_yield(*)` instead of `yield *`, and remain functionally identical, although it would get a performance hit from OS thread switching versus coroutines.

**Comment by ktbarrett** [comment]
Closed by #1171.
